### PR TITLE
dont prompt for immutable params

### PR DESCRIPTION
**Pull request by joehan**
### Description
Adds support for immutable params during ext:configure.
 Instead of prompting for immutable params, we show the user the current value. Since some official extensions don't have LOCATION set, I kept the special casing for it and added a todo to remove it when it is not needed.
	 
### Scenarios Tested
ext:configure
Screen Shot 2019-11-13 at 2 47 58 PM

### CODE REVIEW

File: src/commands/ext-configure.ts [hunk 1]
```diff
@@ -48,21 +48,27 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+
       const params = await paramHelper.getParams(
         projectId,
         paramSpecWithNewDefaults,
         options.params
       );
-      if (removedLocations.length) {
+      if (immutableParams.length) {
+        let immutableParamsText = "";
+        immutableParams.forEach((immutableParam) => {
```
Comment (bkendall) [inline src/commands/ext-configure.ts]: Prefer:

```typescript
for (const immutableParam of immutableParams) {}
```

File: src/commands/ext-configure.ts [hunk 2]
```diff
@@ -48,21 +48,27 @@ export default new Command("ext:configure <extensionInstanceId>")
...
       });
-      const currentLocation = _.get(existingInstance, "config.params.LOCATION");
+
       const params = await paramHelper.getParams(
         projectId,
         paramSpecWithNewDefaults,
         options.params
       );
-      if (removedLocations.length) {
+      if (immutableParams.length) {
```
Comment (bkendall) [inline src/commands/ext-configure.ts]: nit: his check doesn't need to be this early.

If you do the loop, and then check and log `if (immutableParamsText) { logger.info... `, that does the same (and reduces the logic in the if)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,5 @@
 * Fixes a bug where the Functions emulator ignored the "host" configuration (#1722)
 * Fixes a bug where the Functions emulator accepted requests to too many paths (#1773)
+* Adds support for immutable params in ext:configure
 * Adds printing of Firebase console links for Firebase Extensions after operations.
-* Updated Firebase Extensions registry address.
\ No newline at end of file
+* Updates Firebase Extensions registry address.
```
Comment (joehan) [inline CHANGELOG.md]: Switching this over to active voice.

File: src/commands/ext-configure.ts [hunk 4]
```diff
@@ -48,21 +48,27 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+      let immutableParamsText = "";
+      for (const immutableParam of immutableParams) {
+        const currentValue = _.get(existingInstance, "config.params." + immutableParam.param);
+        immutableParamsText += `param: ${immutableParam.param}, value: ${currentValue}\n`;
+        params[immutableParam.param] = currentValue;
+      };
+      if (immutableParamsText) {
         logger.info(
-          `Location is currently set to ${currentLocation}. This cannot be modified. ` +
-            `Please uninstall and reinstall this extension to change location.`
```
Comment (rachelsaunders) [inline src/commands/ext-configure.ts]: I'd change this line to:

"To set a different location, uninstall this extension, then install a new instance of this extension."

File: src/commands/ext-configure.ts [hunk 5]
```diff
@@ -48,21 +48,27 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+          `The following params are immutable:\n${immutableParamsText}These cannot be modified. ` +
+            `Please uninstall and reinstall this extension to change these values.`
```
Comment (rachelsaunders) [inline src/commands/ext-configure.ts]: I'd change this line to:

"To set different values for these params, uninstall this extension, then install a new instance of this extension."

File: src/commands/ext-configure.ts [hunk 6]
```diff
@@ -48,21 +48,27 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+      for (const immutableParam of immutableParams) {
+        const currentValue = _.get(existingInstance, "config.params." + immutableParam.param);
+        immutableParamsText += `param: ${immutableParam.param}, value: ${currentValue}\n`;
+        params[immutableParam.param] = currentValue;
+      };
+      if (immutableParamsText) {
         logger.info(
-          `Location is currently set to ${currentLocation}. This cannot be modified. ` +
-            `Please uninstall and reinstall this extension to change location.`
+          `The following params are immutable:\n${immutableParamsText}These cannot be modified. ` +
```
Comment (rachelsaunders) [inline src/commands/ext-configure.ts]: Looks like there's a missing period and space between the param placeholder and the "These".

Also, quick nit question- does this code/language take into account if only ONE param is in the lists of immutables?  (plural phrasing vs singular phrasing)

Comment (joehan) [reply src/commands/ext-configure.ts]: To the missing period/space: ${immutableParamsText} always ends in '/n' if this is printed, so theres no need - check out the picture in the description to see what it ends up looking like

To the question about plurals: It doesn't right now, but its pretty easy to make it correct, so ill switch it over

Comment (rachelsaunders) [reply src/commands/ext-configure.ts]: Ah!  mis-read the combo of sentences - sorry!  Yes, this is fine to just break the line and not have a period.

Quick fix request: the "To set a different value for this params..." should be "To set a different value for this param..."

File: src/commands/ext-configure.ts [hunk 7]
```diff
@@ -48,21 +48,33 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+
       const params = await paramHelper.getParams(
         projectId,
         paramSpecWithNewDefaults,
         options.params
       );
-      if (removedLocations.length) {
+      let immutableParamsText = "";
+      for (const immutableParam of immutableParams) {
+        const currentValue = _.get(existingInstance, "config.params." + immutableParam.param);
```
Comment (bkendall) [inline src/commands/ext-configure.ts]: Prefer template strings

```suggestion
        const currentValue = _.get(existingInstance, `config.params.${immutableParam.param}`);
```

File: src/commands/ext-configure.ts [hunk 8]
```diff
@@ -48,21 +48,33 @@ export default new Command("ext:configure <extensionInstanceId>")
...
+      }
+      if (immutableParams.length > 1) {
         logger.info(
-          `Location is currently set to ${currentLocation}. This cannot be modified. ` +
-            `Please uninstall and reinstall this extension to change location.`
+          `The following params are immutable:\n${immutableParamsText}These cannot be modified. ` +
+            "To set different values for these params, uninstall this extension, then install a new instance of this extension."
         );
-        params.LOCATION = currentLocation;
       }
```
Comment (bkendall) [inline src/commands/ext-configure.ts]: What would you think of the following? (I tend to not do "build the string" patterns if they can be avoided):

```typescript
const plural = immutableParams.length > 1;
logger.info(`The following param${plural ? "s are" : " is"} immutable:`);
for (const { param } of immutableParams) {
  logger.info(`param: ${param}, value: ${_.get(existingInstance, `config.params.${param}`)}`)
}
// Easier to do one `plural` check here to change the first half of the statement.
logger.info((plural
  ? "To set different values for these params"
  : "To set a different value for this param")
    + ", uninstall the extension, then install a new instance of this extension.");
}
```

This way, the reader isn't trying to track down where newlines are coming from, each param line is generated right in place, it reads top-to-bottom as the logs would appear in the console - I generally would prefer this pattern slightly more. What do you think?

Comment (rachelsaunders) [reply src/commands/ext-configure.ts]: The copy looks fine to me.  Y'all's call on the syntax of making the copy display.

Comment (joehan) [reply src/commands/ext-configure.ts]: I can get behind this - Rachel didn't see the newlines in the syntax I had on first look, so its definitely not the most readable atm.

**Pull request merged**